### PR TITLE
Improve pretty printing

### PR DESCRIPTION
**Pull request by turbolent**
Work towards #209
 
## Description

Finish the implementation of pretty printing all AST elements:
  - Function blocks with pre/post conditions
  - Function expressions and declarations
  - Pragmas
  - Variable declarations, including optional second value
  - Imports
  - Composite declarations, including fields and enum cases
  - Interface declarations
  - Transactions
  - Programs
  - Expressions: Parenthesize sub-expressions when needed

Improve parsing:
- Remove the empty parameter lists from parsed transaction declaration's `execute` "blocks" (special functions)
- Make logical and and logical or left associative. There is no reason to make them right associative (likely an oversight; does not change semantics)

Also, improve the layout of the pretty printing tool (`tools/prettier`).

Give it a try with `go run ./tools/pretty`!

______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

**Review by robert-e-davidson3 — APPROVED** [review]

**Review by SupunS — COMMENTED** [review]
Thats a one big PR  😅  I mostly reviewed the pattern and general go code, and Looks good!

I have two suggestions:
 - One suggestion for test: perhaps add a test-case where a complex program that covers all (or an extensive set of) AST nodes, is parsed from the source and pretty printed, and compared it with the original source. This can eliminate any potential mistakes in manually asserting the "docs". The test can be of two levels:
    - Level 1: Original source code is properly formatted. The round trip should produce the original source code. 
    - Level 2: Further extend above by randomly increasing/changing space/newlines, and check the whether the round trip produces the above properly formatted source.
 
 - It might be good to have the pretty printer code separated from the AST. i.e: make it a visitor (in a separate PR of course).

**Comment by turbolent** [comment]
@SupunS 

> Thats a one big PR  😅  I mostly reviewed the pattern and general go code, and Looks good!
> 
> I have two suggestions:
>  - One suggestion for test: perhaps add a test-case where a complex program that covers all (or an extensive set of) AST nodes, is parsed from the source and pretty printed, and compared it with the original source. This can eliminate any potential mistakes in manually asserting the "docs". The test can be of two levels:
>     - Level 1: Original source code is properly formatted. The round trip should produce the original source code. 
>     - Level 2: Further extend above by randomly increasing/changing space/newlines, and check the whether the round trip produces the above properly formatted source.
>  
>  - It might be good to have the pretty printer code separated from the AST. i.e: make it a visitor (in a separate PR of course). 

Sorry about the size 😅 It's a lot of tests, maybe should have split it up, but kept on hacking away.

These are both two great suggestions. As this PR only improves, but does not yet fully complete the pretty printing implementation, I'll add such a test in a follow up. Also, as this PR is mostly improving on the existing functionality / structure, I'll open a refactor in a separate follow-up PR

**Comment by SupunS** [comment]
yeah, of course!

**Review by SupunS — APPROVED** [review]

File: runtime/ast/block.go [hunk 1]
```diff
@@ -162,6 +160,61 @@ func (b *FunctionBlock) EndPosition(common.MemoryGauge) Position {
...
+	if b.IsEmpty() {
+		return blockEmptyDoc
+	}
+
+	var conditionDocs []prettier.Doc
+
+	if conditionsDoc := b.PreConditions.Doc(preConditionsKeywordDoc); conditionsDoc != nil {
+		conditionDocs = append(
+			conditionDocs,
+			prettier.HardLine{},
```
[the remark this answers is not in the document]
Comment (turbolent) [reply runtime/ast/block.go:188]: All documents are immutable, so they won't get updated in downstream code / inside the prettier library.
It's an empty struct, so IDK if there's any improvement other than saving two characters

File: runtime/ast/type.go [hunk 2]
```diff
@@ -152,7 +144,19 @@ func (t *NominalType) EndPosition(memoryGauge common.MemoryGauge) Position {
...
 
 func (t *NominalType) Doc() prettier.Doc {
-	return prettier.Text(t.String())
+	var doc prettier.Doc = prettier.Text(t.Identifier.String())
+	if len(t.NestedIdentifiers) > 0 {
+		concat := prettier.Concat{doc}
+		for _, identifier := range t.NestedIdentifiers {
+			concat = append(
+				concat,
+				prettier.Text("."),
```
[the remark this answers is not in the document]
Comment (turbolent) [reply runtime/ast/type.go]: Refactored in 0dbc209fef55223eb35538f1f7e1d326f8f282ba

**Pull request merged**